### ISSUE TITLE
Support GraalVM 21.2 based on OpenJDK 16 for native binary

### ISSUE THREAD

**Issue by aminnasiri** [quarkusio/quarkus#19737]
### Description

As a developer, I would like convert my Quarkus application base on OpenJDK 16 to native binary.
This application is using some feature of JDK 16 like record.

```java
public record Vehicle(String brand, String licensePlate) {
    public Vehicle(String brand) {
        this(brand, null);
    }
}
```

**Comment by famod** [comment]
Have you tried Quarkus 2.2.0.CR1 (or 2.2.0.Final, not yet released officially, but already at Maven Central)?

**Comment by aminnasiri** [comment]
Hi Falko,

Yes, I do test version 2.2.1.Final, and this is an error I am getting, it is the same error from version 2.1.3.Final.

```[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildRunner] docker run --env LANG=C --rm -v /Users/amin/Developments/appsrc/practise/aws-lambda/target/aws-lambda-1.0-SNAPSHOT-native-image-source-jar:/project:z --name build-native-AzYXo quay.io/quarkus/ubi-quarkus-native-image:21.2-java16 -J-Djava.util.logging.manager=org.jboss.logmanager.LogManager -J-Dsun.nio.ch.maxUpdateArraySize=100 -J-Dvertx.logger-delegate-factory-class-name=io.quarkus.vertx.core.runtime.VertxLogDelegateFactory -J-Dvertx.disableDnsResolver=true -J-Dio.netty.leakDetection.level=DISABLED -J-Dio.netty.allocator.maxOrder=3 -J-Duser.language=en -J-Duser.country=CA -J-Dfile.encoding=UTF-8 -H:InitialCollectionPolicy=com.oracle.svm.core.genscavenge.CollectionPolicy\$BySpaceAndTime -H:+JNI -H:+AllowFoldMethods -jar aws-lambda-1.0-SNAPSHOT-runner.jar -H:FallbackThreshold=0 -H:+ReportExceptionStackTraces -H:-AddAllCharsets -H:EnableURLProtocols=http,https -H:-UseServiceLoaderFeature -H:+StackTrace -H:-ParseOnce aws-lambda-1.0-SNAPSHOT-runner
[aws-lambda-1.0-SNAPSHOT-runner:28]    classlist:   8,804.72 ms,  0.94 GB
[aws-lambda-1.0-SNAPSHOT-runner:28]        setup:     566.22 ms,  0.93 GB
Error: Annotated class must be final: class io.quarkus.amazon.lambda.http.graal.LambdaContainerHandlerSubstitution
com.oracle.svm.core.util.UserError$UserException: Annotated class must be final: class io.quarkus.amazon.lambda.http.graal.LambdaContainerHandlerSubstitution
        at com.oracle.svm.core.util.UserError.abort(UserError.java:68)
        at com.oracle.svm.core.util.UserError.guarantee(UserError.java:96)
        at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.handleClass(AnnotationSubstitutionProcessor.java:280)
        at com.oracle.svm.hosted.substitute.AnnotationSubstitutionProcessor.init(AnnotationSubstitutionProcessor.java:266)
        at com.oracle.svm.hosted.NativeImageGenerator.createDeclarativeSubstitutionProcessor(NativeImageGenerator.java:936)
        at com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:868)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:530)
        at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:491)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:380)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:543)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:119)
        at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:573)
[aws-lambda-1.0-SNAPSHOT-runner:28]      [total]:   9,415.44 ms,  0.93 GB
# Printing build artifacts to: /project/aws-lambda-1.0-SNAPSHOT-runner.build_artifacts.txt
Error: Image build request failed with exit status 1
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  46.103 s
[INFO] Finished at: 2021-08-31T20:06:21-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:2.2.1.Final:build (default) on project aws-lambda: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: java.lang.RuntimeException: Failed to build native image
[ERROR]         at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:235)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[ERROR]         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
[ERROR]         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[ERROR]         at java.base/java.lang.reflect.Method.invoke(Method.java:567)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:820)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:831)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.lang.RuntimeException: Image generation failed. Exit code: 1
[ERROR]         at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.imageGenerationFailed(NativeImageBuildStep.java:377)
[ERROR]         at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:205)
[ERROR]         ... 11 more
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by kjq** [comment]
I just saw this, does this mean I can drop JDK11 and go towards the latest JDK in the 2.2.x version of Quarkus?

**Comment by aminnasiri** [comment]
@kjq My Quarkus based applications are working fine with JDK16 on the JVM platform, but I can't convert them to a native binary.

**Comment by leozin** [comment]
Yeah I'd say the only issue is the support of JDK16 Records. There are some constraints when it comes to creating a native image and I imagine some reflection procedures are not yet allowed.  

When I try to build a JAX-RS endpoint like this:

```
@Path("/")
public class AuthenticationResource {

    @GET
    @Path("signup")
    public MyRecord signup() {
        return new MyRecord("nay");
    }

    public record MyRecord(String msg) {

    }
}
```
It fails. However, if I just change to:
```
@Path("/")
public class AuthenticationResource {

    @GET
    @Path("signup")
    public String signup() {
        return "aye";
    }

}
```
then the native image is successful.

Will it be possible to have Java Records in native images?

**Comment by jaikiran** [comment]
> Error: Annotated class must be final: class io.quarkus.amazon.lambda.http.graal.LambdaContainerHandlerSubstitution
com.oracle.svm.core.util.UserError$UserException: Annotated class must be final: class io.quarkus.amazon.lambda.http.graal.LambdaContainerHandlerSubstitution
        at com.oracle.svm.core.util.UserError.abort(UserError.java:68)

This error has been fixed as part of https://github.com/quarkusio/quarkus/pull/19887 so that should be available in the already released 2.2.2.Final (and subsequent releases).

**Comment by geoand** [comment]
Closing this in favor of https://github.com/quarkusio/quarkus/issues/20891 which is more up to date.
